src/smc-webapp/smc_chat.cjsx

```diff
@@ -323,27 +323,86 @@ Message = rclass
                 {cols}
             </Row>
 
+sortCopy = (arr) ->
+    return arr.slice(0).sort()
+
+are_dicts_equal = (a, b) ->
+    # assuming maps to lists
+    sorteda = {}
+    for key in sortCopy(Object.keys(a))
+        sorteda[key] = a[key]
+    sortedb = {}
+    for key in sortCopy(Object.keys(b))
+        sortedb[key] = b[key]
+    return JSON.stringify(sorteda) == JSON.stringify(sortedb)
+
+diffArray = (arr1, arr2) ->
+    newArr = []
+    arr1.forEach (val) ->
+        if arr2.indexOf(val) < 0
+            newArr.push val
+        return
+    arr2.forEach (val) ->
+        if arr1.indexOf(val) < 0
+            newArr.push val
+        return
+    newArr
+
 ChatLog = rclass
     displayName: "ChatLog"
 
     propTypes:
         messages     : rtypes.object.isRequired   # immutable js map {timestamps} --> message.
         user_map     : rtypes.object              # immutable js map {collaborators} --> account info
+        read_dates_by_user : rtypes.object
+        message_dates: rtypes.array
         account_id   : rtypes.string
         project_id   : rtypes.string   # optional -- used to render links more effectively
         file_path    : rtypes.string   # optional -- ...
+        path         : rtypes.string
         font_size    : rtypes.number
         actions      : rtypes.object
         show_heads   : rtypes.bool
         focus_end    : rtypes.func
         saved_mesg   : rtypes.string
         set_scroll   : rtypes.func
 
+    getInitialState: ->
+        read_indexes_to_user_ids: @find_read_tos(@props.read_dates_by_user, @props.message_dates)
+        prev_read_dates_by_user: @props.read_dates_by_user
+
+    mixins: [SetIntervalMixin]
+
+    componentWillMount: ->
+        setInterval(@update_read_indexes_to_user_ids, 300)
+
+    update_read_indexes_to_user_ids: ->
+        if are_dicts_equal(@props.read_dates_by_user, @state.prev_read_dates_by_user)
+            return
+        read_indexes_to_user_ids = @state.read_indexes_to_user_ids
+        # figure out which users to update
+        diff = diffArray(Object.keys(@props.read_dates_by_user), Object.keys(@state.prev_read_dates_by_user))
```

Comment:
diffArray does not work on associative arrays which is what `@props.read_dates_by_user, @state.prev_read_dates_by_user` are